### PR TITLE
fix: HELLO AUTH with non default user

### LINKED ISSUE #3476: Inability to Authorize Non-Default User with RESP3 Protocol by `HELLO 3 AUTH non_default_user some_pass`
**Describe the bug**
When attempting to connect using a non-default user and the RESP3 protocol `HELLO 3 AUTH non_default_user some_pass`, an error message is returned: "WRONGPASS invalid username-password pair or user is disabled." However, the authorization succeeds when using the `AUTH` command directly without the RESP3 protocol by `AUTH non_default_user some_pass` but then using RESP2.

**To Reproduce**
Attempt to connect as a non-default user with the RESP3 protocol
```
> HELLO 3 AUTH non_default_user some_pass
> (error) WRONGPASS invalid username-password pair or user is disabled.
```

Attempt to authenticate using the non-default user without the RESP3 protocol:
```
> AUTH non_default_user some_pass
> OK
```

Attempt to connect using the RESP3 protocol with the default user:
```
> HELLO 3 AUTH default some_default_pass
> 1# "server" => "redis"
...
```

**Expected behavior**
The non-default user should be authorized when connecting with the RESP3 protocol without encountering the WRONGPASS error as `HELLO 3 AUTH non_default_user some_pass`.

**Environment (please complete the following information):**
 - OS: 22.04.1-Ubuntu SMP
 - Kernel: 6.5.0-14-generic 
 - Containerized: Docker (docker.dragonflydb.io/dragonflydb/dragonfly:1.21.0)
 - Dragonfly Version: df-v1.21.0

**Reproducible Code Snippet**
```
# Attempt to connect with RESP3 protocol and non-default user
HELLO 3 AUTH non_default_user some_pass
```

**Additional context**
In the source code, when using AUTH with a non-default user, the following error is sent: SendError(facade::kAuthRejected).
Reference: https://github.com/dragonflydb/dragonfly/blob/7b9bbc19d7df29a7b8651ed9060eb11ef883d5e0/src/server/server_family.cc#L2421
It's unclear why this behavior occurs and why documentation doesn't mention this limitation.

### PR DESCRIPTION
**Pull request by kostasrim**
fixes #3476 

* fix HELLO AUTH with non default user
* add test_auth_resp3_bug
* add decode_responses=True to acl tests and refactor
* fix test_default_user_bug

### CODE REVIEW

File: src/server/dragonfly_test.cc [hunk 1]
```diff
@@ -316,9 +324,14 @@ TEST_F(DflyEngineTest, Hello) {
               ErrArg("WRONGPASS invalid username-password pair or user is disabled."));
 
   resp = Run({"hello", "3", "AUTH", "default", ""});
+  ASSERT_THAT(resp, ErrArg("WRONGPASS invalid username-password pair or user is disabled."));
```
Comment (kostasrim) [inline src/server/dragonfly_test.cc:327]: Valkey does not allow AUTH user without password, now we got the correct semantics.

File: src/server/server_family.cc [hunk 2]
```diff
@@ -1626,38 +1626,49 @@ void ServerFamily::FlushAll(CmdArgList args, ConnectionContext* cntx) {
   cntx->SendOk();
 }
 
+bool ServerFamily::DoAuth(ConnectionContext* cntx, std::string_view username,
+                          std::string_view password) const {
+  const auto* registry = ServerState::tlocal()->user_registry;
+  const bool is_authorized = registry->AuthUser(username, password);
```
Comment (chakaz) [inline src/server/server_family.cc:1632]: Add `CHECK(registry != nullptr);`?

File: src/server/server_family.cc [hunk 3]
```diff
@@ -1626,38 +1626,49 @@ void ServerFamily::FlushAll(CmdArgList args, ConnectionContext* cntx) {
...
+    cntx->authenticated = true;
+  }
+  return is_authorized;
+}
+
 void ServerFamily::Auth(CmdArgList args, ConnectionContext* cntx) {
   if (args.size() > 2) {
-    return cntx->SendError(kSyntaxErr);
+    cntx->SendError(kSyntaxErr);
+    return;
```
Comment (chakaz) [inline src/server/server_family.cc:1647]: Why change this? We do it in many places, it's more convenient and saves some verbosity..

Comment (chakaz) [reply src/server/server_family.cc:1647]: (same below in many places)

File: tests/dragonfly/acl_family_test.py [hunk 4]
```diff
@@ -135,51 +135,51 @@ async def test_acl_cat_commands_multi_exec_squash(df_factory):
     df.start()
 
     # Testing acl categories
-    client = aioredis.Redis(port=df.port)
+    client = aioredis.Redis(port=df.port, decode_responses=True)
```
Comment (chakaz) [inline tests/dragonfly/acl_family_test.py:138]: Nice!

**Review by chakaz — COMMENTED** [review]
Overall looks good, please see my minor comments

File: src/server/server_family.cc (hunk 3 above)
Comment (kostasrim) [reply src/server/server_family.cc:1647]: well IMO I find `return <<void expression>>` confusing but I am happy to revert this if you find that it's confusing

Comment (kostasrim) [reply src/server/server_family.cc:1647]: will change this soon

Comment (chakaz) [reply src/server/server_family.cc:1647]: Yes, please revert :) Thanks!

Comment (kostasrim) [reply src/server/server_family.cc]: done

File: src/server/server_family.cc (hunk 2 above)
Comment (kostasrim) [reply src/server/server_family.cc:1633]: done, but with `CHECK(registry)` :nerd_face:

Comment (chakaz) [reply src/server/server_family.cc:1633]: I think that comparing to `nullptr` is a clearer phrasing, but :shrug:

**Review by chakaz — APPROVED** [review]

**Pull request merged**